sway-lib-std/src/vm/evm/ecr.sw

```diff
@@ -0,0 +1,51 @@
+library ecr;
+
+use ::address::Address;
+use ::b512::B512;
+use ::context::registers::error;
+use ::hash::{HashMethod, hash_pair};
+use ::result::*;
+
+pub enum EcRecoverError {
+    UnrecoverablePublicKey: (),
+}
+
+/// Recover the public key derived from the private key used to sign a message.
+/// Returns a `Result` to let the caller choose an error handling strategy.
+pub fn ec_recover(signature: B512, msg_hash: b256) -> Result<B512, EcRecoverError> {
+    let public_key = ~B512::new();
+    let was_error = asm(buffer: public_key.bytes, sig: signature.bytes, hash: msg_hash) {
+        ecr buffer sig hash;
+        err
+    };
+    // check the $err register to see if the `ecr` opcode succeeded
+    if was_error == 1 {
+        Result::Err(EcRecoverError::UnrecoverablePublicKey)
+    } else {
+        Result::Ok(public_key)
+    }
+}
+
+/// Recover the address derived from the private key used to sign a message.
+/// Returns a `Result` to let the caller choose an error handling strategy.
+/// Ethereum addresses are 20 bytes long, so these are left-padded to fit in a 32 byte Address type.
+pub fn ec_recover_address(signature: B512, msg_hash: b256) -> Result<Address, EcRecoverError> {
+    let pub_key_result = ec_recover(signature, msg_hash);
+
+    if let Result::Err(e) = pub_key_result {
+        // propagate the error if it exists
+        Result::Err(e)
```

Comment:
Instead of this, maybe just return `pub_key_result` if `is_error()`?

Reply:
I believe this suggestion won't work. The types are different even though they appear to line up.

Reply:
Oh I see because `EcRecoverError` is being re-declared. In that case, just don't re-declare it and use the one in that already exists!

Reply:
Yes this is all because I started with a copy paste of ecr.sw and forgot to tidy up the redundancies. Only the 2nd function here needs to exist.

Reply:
> banishes half the file to the shadow realm

Reply:
hmm the types don't line up though. `ec_recover_address` returns `Result<Address, EcRecoverError>`, whereas `ec_recover` returns `Result<B512, EcRecoverError>`

Even with the syntax as it is, i'm getting some weirdness where with `ec_recover` copy/pasted, it works, but when I import it from `std::ecr`, I get this : 

` let recovered_address = result.unwrap();
   |                                    ^^^^^^ No method named "unwrap" found for type "enum Result<struct Address<b256>, enum EcRecoverError<()>>".`

Reply:
> hmm the types don't line up though.

Ah I see. Well you still should be able to re-use the `EcRecoverError` from `ecr` at least.

Reply:
Resolved by [39e4fba](https://github.com/FuelLabs/sway/pull/1374/commits/39e4fbaaa3481b8acbd05db4744cbb2d4b5f2073)